File: frontend/src/api/transportCosts.ts
import type { RecommendationResponse, TransportType } from './recommendation'

export type TransportCostRecord = {
  id: number
  travel_date: string
  actual_cost_won: number
  recommended_cost_won: number
  potential_savings_won: number
  selected_transport_type: TransportType
  recommended_transport_type: TransportType
  created_at: string
}

export type MonthlyTransportCostResponse = {
  month: string
  daily_summaries: Array<{
    date: string
    record_count: number
    actual_cost_won: number
    recommended_cost_won: number
    potential_savings_won: number
  }>
  totals: {
    actual_cost_won: number
    recommended_cost_won: number
    potential_savings_won: number
  }
}

export type DailyTransportCostResponse = {
  date: string
  records: TransportCostRecord[]
  totals: {
    actual_cost_won: number
    recommended_cost_won: number
    potential_savings_won: number
  }
}

const apiBaseUrl = (import.meta.env.VITE_API_BASE_URL as string | undefined)?.replace(/\/$/, '')
  || 'http://127.0.0.1:8000'

export async function createTransportCostRecord(
  result: RecommendationResponse,
  travelDate: string,
  actualCostWon: number,
  selectedTransportType: TransportType,
): Promise<TransportCostRecord> {
  const response = await fetch(`${apiBaseUrl}/transport-cost-records`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      travel_date: travelDate,
      actual_cost_won: actualCostWon,
      selected_transport_type: selectedTransportType,
      origin: result.origin,
      destination: result.destination,
      transport_types: result.transport_types,
    }),
  })
  if (!response.ok) throw new Error(`Transport cost request failed with status ${response.status}`)
  return (await response.json()) as TransportCostRecord
}

export async function fetchMonthlyTransportCosts(
  month: string,
  signal?: AbortSignal,
): Promise<MonthlyTransportCostResponse> {
  const response = await fetch(
    `${apiBaseUrl}/transport-cost-records/monthly?month=${encodeURIComponent(month)}`,
    { signal },
  )
  if (!response.ok) throw new Error(`Monthly transport cost request failed with status ${response.status}`)
  return (await response.json()) as MonthlyTransportCostResponse
}

export async function fetchDailyTransportCosts(
  date: string,
  signal?: AbortSignal,
): Promise<DailyTransportCostResponse> {
  const response = await fetch(
    `${apiBaseUrl}/transport-cost-records/daily?date=${encodeURIComponent(date)}`,
    { signal },
  )
  if (!response.ok) throw new Error(`Daily transport cost request failed with status ${response.status}`)
  return (await response.json()) as DailyTransportCostResponse
}

File: backend/app/api/transport_costs.py
"""교통비 기록 생성과 날짜·월별 조회 API."""

from datetime import date

from fastapi import APIRouter, Depends, HTTPException, Query, status

from app.api.contracts import (
    DailyTransportCostResponse,
    MonthlyTransportCostResponse,
    RecommendationPriority,
    TransportCostRecord,
    TransportCostRecordCreate,
)
from app.api.recommendation import get_recommendation_route_provider
from app.core.config import get_settings
from app.services.recommendation import RecommendationRouteProvider, rank_routes
from app.services.transport_costs import SqliteTransportCostRepository, TransportCostRepositoryError

router = APIRouter(prefix="/transport-cost-records", tags=["transport-costs"])


def get_transport_cost_repository() -> SqliteTransportCostRepository:
    return SqliteTransportCostRepository(get_settings().resolved_transport_cost_db_path)


@router.post("", response_model=TransportCostRecord, status_code=status.HTTP_201_CREATED)
def create_transport_cost_record(
    request: TransportCostRecordCreate,
    route_provider: RecommendationRouteProvider = Depends(get_recommendation_route_provider),
    repository: SqliteTransportCostRepository = Depends(get_transport_cost_repository),
) -> TransportCostRecord:
    routes = route_provider.get_routes(request.origin, request.destination, request.transport_types)
    recommendations, _ = rank_routes(
        routes,
        [RecommendationPriority.COST, RecommendationPriority.TIME, RecommendationPriority.WALK],
    )
    if not recommendations or recommendations[0].route.total_cost_won is None:
        raise HTTPException(status_code=503, detail="금액 우선 추천 비용을 계산할 수 없습니다.")
    recommended_route = recommendations[0].route
    try:
        return repository.add(
            travel_date=request.travel_date,
            actual_cost_won=request.actual_cost_won,
            recommended_cost_won=recommended_route.total_cost_won,
            selected_transport_type=request.selected_transport_type,
            recommended_transport_type=recommended_route.transport_type,
            origin=request.origin,
            destination=request.destination,
        )
    except TransportCostRepositoryError as exc:
        raise HTTPException(status_code=503, detail=str(exc)) from exc


@router.get("/daily", response_model=DailyTransportCostResponse)
def get_daily_transport_costs(
    date_: date = Query(alias="date"),
    repository: SqliteTransportCostRepository = Depends(get_transport_cost_repository),
) -> DailyTransportCostResponse:
    try:
        return repository.daily(date_)
    except TransportCostRepositoryError as exc:
        raise HTTPException(status_code=503, detail=str(exc)) from exc


@router.get("/monthly", response_model=MonthlyTransportCostResponse)
def get_monthly_transport_costs(
    month: str = Query(pattern=r"^\d{4}-(0[1-9]|1[0-2])$"),
    repository: SqliteTransportCostRepository = Depends(get_transport_cost_repository),
) -> MonthlyTransportCostResponse:
    try:
        return repository.monthly(month)
    except TransportCostRepositoryError as exc:
        raise HTTPException(status_code=503, detail=str(exc)) from exc
